### PR TITLE
[Resolves #1213] Add shell option to cmd hook

### LINKED ISSUE #1213: !cmd hook runs in `sh`, which is not my default shell
### Subject of the issue
When I use a `!cmd` hook, it appears to be run by `sh` because I receive this error: 

```shell
/bin/sh: 1: [[: not found
```

Is there a way to force the underlying `subprocess.run()` call to use Python? I know I can do this in Python by using arguments that look like `["bash", "-c", "your command goes here"]`, and, in fact, if I simply pass in `["your command goes here"]` in Python, it will run in bash because it is my default shell. Is `sceptre` really running my hook using `sh`? If so, how can I make it use `bash` instead?

### Your environment
* version of sceptre (sceptre --version) 
**Sceptre, version 2.4.0**
* version of python (python --version) 
**Python 3.8.10**
* which OS/distro
**Ubuntu 20.04.3 LTS on 4.4.0-19041-Microsoft #1237-Microsoft x86_64** 

### Steps to reproduce
Tell us how to reproduce this issue. Please provide sceptre project files if possible,
you can use https://plnkr.co/edit/ANFHm61Ilt4mQVgF as a base.

Here is a hook that causes the behavior:
```yaml
hooks:
    before_update:
        - !cmd 'if [[ "{{ var.update_lambdas | default("false") }}" == "true" ]]; then lambdas/deploy-lambdas.sh -u "$(./get-stack-output.sh "{{ environment }}/{{ region }}/LambdaS3Bucket.yaml" LambdaBucketName)" HandleServiceAccountAppGitHubWebhook; fi;'
```

### Expected behaviour
The hook should run successfully using my default shell, which is `bash`.

### Actual behaviour
I see the error `/bin/sh: 1: [[: not found` instead.

### PR DESCRIPTION
**Pull request by iainelder**
The cmd hook now accepts named arguments `run` and `shell` to run a command in a
given shell. For backwards compatibility it still supports the unnamed
argument to run a command in the default shell.

The documentation shows the expected output for examples of each argument style.

The unit tests no longer depend on the implementation of Cmd and test only the
behavior. This makes it easier to change how the hook controls the subprocess.

## PR Checklist

- [x] Wrote a good commit message & description [see guide below].
- [x] ~Commit message~ PR title starts with `[Resolve #issue-number]`.
- [x] Added/Updated unit tests.
- [ ] Added/Updated integration tests (if applicable).
- [x] All unit tests (`poetry run tox`) are passing.
- [x] Used the same coding conventions as the rest of the project.
- [x] The new code passes pre-commit validations (`poetry run pre-commit run --all-files`).
- [x] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

### CODE REVIEW

File: sceptre/hooks/cmd.py [hunk 1]
```diff
@@ -19,8 +19,16 @@ def run(self):
...
+            executable = None
+        elif isinstance(self.argument, dict):
+            args = self.argument["args"]
+            executable = self.argument["executable"]
+
         try:
-            subprocess.check_call(self.argument, shell=True, env=envs)
+            subprocess.check_call(args, shell=True, env=envs, executable=executable)
         except TypeError:
             raise InvalidHookArgumentTypeError(
```
Comment (iainelder) [inline sceptre/hooks/cmd.py:33]: The error message needs to change to reflect the new options. And what type of inputs will be invalid now? I hope after I learn about the unit tests it will be clear how to rewrite this.

**Comment by iainelder** [comment]
See my [sceptre-dev-notes](https://github.com/iainelder/sceptre-dev-notes/blob/main/dev-notes/cmd-hook-options.md) repo for the notes on the devleopment and testing. If any of that is useful to put in the main repo I can add it here.

**Comment by iainelder** [comment]
Well, tox gives me some failing tests, so I guess I should start here!

```text
Results (15.28s):
     635 passed
       2 failed
         - tests/test_hooks/test_cmd.py:18 TestCmd.test_run_with_non_str_argument
         - tests/test_hooks/test_cmd.py:23 TestCmd.test_run_with_str_argument
ERROR: InvocationError for command /home/isme/.local/bin/poetry run pytest --cov=sceptre --cov-append --cov-report=term-missing --junitxml=test-reports/junit-py310.xml (exited with code 1)
```

File: sceptre/hooks/cmd.py [hunk 2]
```diff
@@ -19,10 +19,15 @@ def run(self):
...
-                "arguments of type string.".format(self.argument)
-            )
+
+        if self.argument is None:
+            raise InvalidHookArgumentTypeError()
+
+        if isinstance(self.argument, str) or self.argument is None:
+            args = self.argument
+            executable = None
+        elif isinstance(self.argument, dict):
```
Comment (zaro0508) [inline sceptre/hooks/cmd.py:29]: passing in a dict is new so could you please add some docs for this?  an example would probably help as well.

also there should probably an else condition?  what if user passes in a list?  or you could probably keep the try/catch to handle all the other wacko things that users do.

File: tests/test_hooks/test_cmd.py [hunk 3]
```diff
@@ -11,27 +10,34 @@
...
         self.cmd = Cmd(stack=self.stack)
 
     def test_run_with_non_str_argument(self):
         self.cmd.argument = None
         with pytest.raises(InvalidHookArgumentTypeError):
             self.cmd.run()
 
-    @patch("sceptre.hooks.cmd.subprocess.check_call")
-    def test_run_with_str_argument(self, mock_call):
+    def test_run_with_str_argument(self, capfd):
```
Comment (zaro0508) [inline tests/test_hooks/test_cmd.py:33]: remember to test running with a dict arg as well.

File: sceptre/hooks/cmd.py [hunk 4]
```diff
@@ -19,10 +19,15 @@ def run(self):
...
-        except TypeError:
-            raise InvalidHookArgumentTypeError(
-                'The argument "{0}" is the wrong type - cmd hooks require '
-                "arguments of type string.".format(self.argument)
-            )
+
+        if self.argument is None:
+            raise InvalidHookArgumentTypeError()
+
+        if isinstance(self.argument, str) or self.argument is None:
```
Comment (zaro0508) [inline sceptre/hooks/cmd.py:26]: doesn't the previous condition already check `self.argument is None`?

**Review by zaro0508 — CHANGES_REQUESTED** [review]

File: sceptre/hooks/cmd.py (hunk 4 above)
Comment (iainelder) [reply sceptre/hooks/cmd.py:26]: Good catch. I added the previous condition but didn't change it here.

File: sceptre/hooks/cmd.py (hunk 2 above)
Comment (iainelder) [reply sceptre/hooks/cmd.py:29]: Indeed, I'll add some tests for invalid inputs so that I can handle those properly.

File: tests/test_hooks/test_cmd.py (hunk 3 above)
Comment (iainelder) [reply tests/test_hooks/test_cmd.py:33]: Yes, I'm looking at how to do that. I've tried to make my tests depend only on the behavior and not the implementation. Just now the test doesn't know that subprocess is being used.

But I don't know how to to test that the `executable` option really causes a different shell to be launched.

I think I might have to use a mock after all and make the tests depend on the `subprocess` module. That module is the obvious way to implement this hook, so maybe it's not so bad.

**Comment by iainelder** [comment]
I've addressed all the feedback about the code. Now I need to add documentation for the new features.

**Comment by iainelder** [comment]
@zaro0508 , how does it look now?

Should I add some tests that run the sceptre CLI on some YAML files? I've just done that testing in my own terminal.

File: docs/_source/docs/hooks.rst [hunk 5]
```diff
@@ -44,23 +44,72 @@ Available Hooks
...
 
-   before_create:
-     - !cmd "echo hello"
+   hooks:
+     before_update:
+       - !cmd
+           command: "if foo; then baz; else qux; fi"
+           shell: "bash"
+
+This example fails because none of the ``foo``, ``baz``, and ``quz`` commands exist. Its standard error messages sit between Sceptre's status messages and a Python traceback.
```
Comment (zaro0508) [inline docs/_source/docs/hooks.rst:101]: seems odd to me to use an example that fails.  i'm not sure I understand what this example is trying to demonstrate specifically?  why not just do  `command: "echo $0"` which should echo "bash"?

**Review by zaro0508 — APPROVED** [review]
LGTM, just had a question in docs.

File: docs/_source/docs/hooks.rst (hunk 5 above)
Comment (iainelder) [reply docs/_source/docs/hooks.rst:101]: Yes, good point. I will split this into two examples to make it clearer.

* one that echoes the shell name and uses Bash syntax to prove we are using the Bash shell
* one that fails (could be any shell) to show what Sceptre does with the shell error

Comment (iainelder) [reply docs/_source/docs/hooks.rst]: @zaro0508 I used your simpler example and added a new one to show the error behavior. I added a PowerShell example too since it's a common choice on Windows. How does it look now?

**Comment by iainelder** [comment]
To avoid confusing people who try this on an old version, I want the new syntax to cause an error in Sceptre v4.2.2.

When `command` is the first named argument, Sceptre v4.2.2 executes `sh -c command shell`. This has no output and returns 0, which makes it look like Sceptre just ignored the hook.

As a compromise I suggest we replace `command` with `run` because it breaks in a more obvious way. The goal here is to avoid confusing people who try the new syntax on an old version of Sceptre.

[GitHub Actions](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsrun) uses `run` to mean the same thing, so we would be in good company.

```pycon
>>> subprocess.check_call({"run": "echo $0", "shell": "bash"}, shell=True)
shell: 1: run: not found
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '{'run': 'echo $0', 'shell': 'bash'}' returned non-zero exit status 127.
```

File: sceptre/hooks/cmd.py [hunk 6]
```diff
@@ -13,16 +13,33 @@ def __init__(self, *args, **kwargs):
...
+            shell = None
+
+        elif (
+            isinstance(self.argument, dict)
+            and set(self.argument) == {"run", "shell"}
+            and isinstance(self.argument["run"], str)
+            and isinstance(self.argument["shell"], str)
+        ):
+            command_to_run = self.argument["run"]
+            shell = self.argument["shell"]
```
Comment (zaro0508) [inline sceptre/hooks/cmd.py:36]: i didn't even consider windows.  What happens if users attempt to use a windows shell on linux or the other way around?  should we bring back the try/catch or add some error handling around `shell` inputs as well?

File: docs/_source/docs/hooks.rst [hunk 7]
```diff
@@ -44,23 +44,105 @@ Available Hooks
...
+Write the executable name as you would type it at an interactive prompt to start the shell. For
+example, if Bash is in the system path, you can write ``bash``; otherwise, you need to write the
+absolute path such as ``/bin/bash``.
+
+.. code-block:: yaml
+
+   hooks:
+     before_update:
+       - !cmd
+           run: echo "Hello, $0!"
```
Comment (zaro0508) [inline docs/_source/docs/hooks.rst:100]: this parameter change is not backwards compatible.  I think this change should be in a separate PR with a separate discussion thread

**Review by zaro0508 — CHANGES_REQUESTED** [review]

File: sceptre/hooks/cmd.py (hunk 6 above)
Comment (iainelder) [reply sceptre/hooks/cmd.py:36]: I'm not against adding a try/catch around the shell invocation. I just don't know what would be useful to catch yet. It might be useful later to wrap the stock Python exception with a Sceptre exception for nicer presentation of the errors, but I had considered that out of scope for now.

I don't think it's our business to know which shells are available in the environment. If the user tells Sceptre to use a shell that isn't available, then Sceptre should fail. I'll add a test and documentation to show what that failure looks like. This is useful to know how it behaves when I do something silly like type `bsah` instead of `bash`.

(PowerShell ships with Windows, but it's no longer just a shell for Windows. I [installed it on Ubuntu](https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-linux?view=powershell-7.3) to test the `pwsh` part of the documentation! Similarly, I believe [Windows for Subsystem for Linux (WSL)](https://learn.microsoft.com/en-us/windows/wsl/install) users can use the Bash shell.)

File: docs/_source/docs/hooks.rst (hunk 7 above)
Comment (iainelder) [reply docs/_source/docs/hooks.rst:100]: Can you tell me what "backwards compatible" means for this feature?

The latest released version of Sceptre is v4.2.2 and supports only the string argument. The keyword arguments, whatever we decide to call them, are a new feature. I thought that as long as we continue to support the string argument then we satisfy the backwards-compatible requirement.

Over the weekend I renamed `command` to `run`, but only in the branch.

Comment (zaro0508) [reply docs/_source/docs/hooks.rst:100]: oops, I misread.  i thought you wanted to replace `!cmd` with `!run`.  sorry about that.  replacing `command` is fine since it's never been used before.

File: sceptre/hooks/cmd.py (hunk 6 above)
Comment (zaro0508) [reply sceptre/hooks/cmd.py:37]: hmm, `FileNotFoundError` for invalid shell name is not very helpful.  would you be able to provide a more helpful message?

**Review by zaro0508 — CHANGES_REQUESTED** [review]

File: docs/_source/docs/hooks.rst (hunk 7 above)
Comment (iainelder) [reply docs/_source/docs/hooks.rst:100]: Ah, okay. Yes, there's no need to change the name of the hook. Apparently that's controlled in poetry.toml.

https://github.com/Sceptre/sceptre/blob/33404ee5486ff345ed738044aeb1cc3b71fdb129/pyproject.toml#L30

File: docs/_source/docs/hooks.rst (hunk 5 above)
Comment (zaro0508) [reply docs/_source/docs/hooks.rst]: looks good

File: sceptre/hooks/cmd.py (hunk 6 above)
Comment (iainelder) [reply sceptre/hooks/cmd.py:37]: The full error message is `FileNotFoundError: [Errno 2]` No such file or directory: '/bin/bsah'`. It does print the bad shell name. But we could improve the message with a Sceptre error that tells the user to check the shell input.

Since we've started talking about custom error handling, it would be great to suppress the stack trace for these kinds of errors that we know how to handle. Currently Sceptre will print about 50 lines of unhelpful stack trace between the first useful error message and the last useful error message. I'll see what I can do about that.

Comment (zaro0508) [reply sceptre/hooks/cmd.py:37]: ohh, the full message looks good to me.  It's up to you but I don't think you need to add supression of the stack trace on this PR.  i'm ok if you want to do it as a separate change.

**Review by zaro0508 — CHANGES_REQUESTED** [review]
could you please update the subject of this PR to include the issue, something like `[Resolves #1213] Add options to cmd hook`?

This branch is out of date so you will need to update (merge) this PR with master.

File: sceptre/hooks/cmd.py (hunk 6 above)
Comment (iainelder) [reply sceptre/hooks/cmd.py:37]: Thanks. I agree about using another PR. The stack trace suppression happens elsewhere in the code. I don't want to make this PR any wider.

**Comment by iainelder** [comment]
I renamed the issue and merged the main branch.

I made the error message checking of the tests more robust.

How does it look now?

**Comment by iainelder** [comment]
I understand that I need to squash all the commits before you merge it. I'll have a look at the other merge commit messages.

**Comment by iainelder** [comment]
I updated the title and description with my proposal for the final commit message.

**Review by zaro0508 — APPROVED** [review]
nice work @iainelder
